### PR TITLE
chore(): Refactor Moderators Query

### PR DESCRIPTION
**Pull request by guivinicius**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description
Change how we query moderators for internal pages.

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x ] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by guivinicius** [comment]
@mstruve can you give me your opinion here? I'm not really happy with what I'm doing for now. Seems odd.

File: app/queries/internal/moderators_query.rb [hunk 1]
```diff
@@ -4,34 +4,21 @@ class ModeratorsQuery
...
-    def self.role_id_for(role)
-      Role.find_by!(name: role).id
-    end
-
-    def self.search_relation(relation, search)
-      relation.where(
-        "users.username ILIKE :search OR users.name ILIKE :search",
-        search: "%#{search}%",
-      )
+      Kaminari.paginate_array(User.find(user_ids), total_count: results.first&.dig("total"))
```
Comment (mstruve) [inline app/queries/internal/moderators_query.rb:21]: You can skip this part and pass the pagination params to our Search class and it will paginate for you

File: spec/support/elasticsearch_helpers.rb [hunk 2]
```diff
@@ -1,7 +1,9 @@
 # Only included in specs with elasticsearch: true
 module ElasticsearchHelpers
   def index_documents(resources)
-    Array.wrap(resources).each(&:index_to_elasticsearch_inline)
-    described_class.refresh_index
+    Array.wrap(resources).each do |resource|
+      resource.index_to_elasticsearch_inline
+      resource::SEARCH_CLASS.refresh_index
```
Comment (mstruve) [inline spec/support/elasticsearch_helpers.rb:6]: There is no need to do this after every index and it will only slow us down. You only need to refresh the index once after everything has been indexed.

File: config/elasticsearch/mappings/users.json [hunk 3]
```diff
@@ -8,6 +8,9 @@
       "type": "keyword",
       "copy_to": "search_fields"
     },
+    "badge_achievements_count": {
```
Comment (mstruve) [inline config/elasticsearch/mappings/users.json:11]: Since we are adding new fields here for the view you will want to add these here, to the serializer, and then run a data update script to reindex all users so the new fields are present. That should all be done in a separate PR from searching.

File: app/queries/internal/moderators_query.rb (hunk 1 above)
Comment (guivinicius) [reply app/queries/internal/moderators_query.rb:21]: Yeah, I'm doing that at line 18.

The issue is that some internal pages use `Kaminari` pagination helper that requires some specific attributes which an array doesn't have.

I rather use the doc results from the `elasticsearch` and not use AR to search again for the records. But for that, I might need to change the `search_documents` results structure for something like:

```ruby
{
  total_count: 100,
  page: 1,
  per_page: 20,
  pages: 5,
  users: [
    { id: 1, name: 'User Name', roles: [] },
    { id: 2, name: 'User Name 2', roles: []}
  ]
}
```

That way we could easily create pagination on front-end using direct `elasticsearch` results and also instantiate User objects from its attributes or just use OpenStruct if we just need to access the values.

```ruby
results = Search::User.search_documents(params: {})
users = results['users'].map |attr|
  OpenStruct.new(attr)
end
```

Something around those lines. WDYT?

Comment (mstruve) [reply app/queries/internal/moderators_query.rb:21]: I like the idea of returning additional metadata with the search documents. I could see that being useful in many places.

**Comment by guivinicius** [comment]
I'm also closing this one for now.
